rclpy/rclpy/executors.py

```diff
@@ -562,6 +562,11 @@ def _wait_for_ready_callbacks(
                         guard_handles.append(gc.handle)
                     except InvalidHandle:
                         entity_count.num_guard_conditions -= 1
+                    except AttributeError:
```

Comment:
From my opinions, I don't think it should throw an exception.     
#968 is possible to occur. So code should be updated to handle this case to avoid exception during shutdown.

Reply:
right, this does not address root cause which is racy condition for `guards`.

Reply:
this PR is under discussing, see https://github.com/ros2/rclpy/issues/968#issuecomment-1196160697 for detail.

Reply:
I actually think this patch makes sense now.

I was considering the following.

```patch
diff --git a/rclpy/rclpy/executors.py b/rclpy/rclpy/executors.py
index 3c5bea4..73b2ba5 100644
--- a/rclpy/rclpy/executors.py
+++ b/rclpy/rclpy/executors.py
@@ -514,8 +514,13 @@ class Executor:
             if timeout_timer is not None:
                 timers.append(timeout_timer)

-            guards.append(self._guard)
-            guards.append(self._sigint_gc)
+            with self._shutdown_lock:
+                if self._guard and self._sigint_gc:
+                    # make sure to append valid GuardConditions
+                    guards.append(self._guard)
+                    guards.append(self._sigint_gc)
+                else:
+                    raise ShutdownException()

             entity_count = NumberOfEntities(
                 len(subscriptions), len(guards), len(timers), len(clients), len(services))
```

but this does not guarantee the valid GuradConditions here,

https://github.com/ros2/rclpy/blob/649bdec92e74eee6ebb45bd2815df9ea9ff48654/rclpy/rclpy/executors.py#L561-L562

Reply:
We think current design is not good.  
Only catching the exception isn't a good solution.   

We want to  
- Executor.shutdown() only set _is_shutdown and notify executor by guard trigger.
- Real shutdown process is moved to Executor.wait_for_ready_callbacks().  

This thought need to be discussed.  A comment is added to https://github.com/ros2/rclpy/issues/968#issuecomment-1196160697